File: solutions/0155-min-stack/solution.cpp
class MinStack {
public:
    vector<pair<int ,int>>s;
    int tp=-1;
    int mn;
    MinStack() {
        
    }
    
    void push(int val) {
        if(tp == -1)
            mn=val;
        else
            mn=min(s[tp].second,val);
        tp++;
        s.push_back({val,mn});
        
    }
    
    void pop() {
        tp--;
        s.pop_back();
    }
    
    int top() {
        return s.back().first;
        
    }
    
    int getMin() {
        return s.back().second;
        
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(val);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: solutions/0155-min-stack/solution.java
class MinStack {
    Stack<Integer>s;
    int a=Integer.MAX_VALUE;
    public MinStack() {
        s=new Stack<>();
    }
    
    public void push(int val) {
        if (val <= a) {
            s.push(a);
            a = val;
        }
        s.push(val);
    }
    
    public void pop() {
        if (s.pop() == a)
            a = s.pop(); 
    }
    
    public int top() {
        return s.peek();
    }
    
    public int getMin() {
        return a;
    }
}

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(val);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */
